reana_server/reana_admin/cli.py

```diff
@@ -716,3 +718,159 @@ def queue_consume(
         logging.exception(error)
     else:
         consumer.run()
+
+
+@reana_admin.command("check-workflows")
+@click.option(
+    "--date-start",
+    type=click.DateTime(formats=["%Y-%m-%d", "%Y-%m-%dT%H:%M:%S"]),
+    default=None,
+)
+@click.option(
+    "--date-end",
+    type=click.DateTime(formats=["%Y-%m-%d", "%Y-%m-%dT%H:%M:%S"]),
+    default=None,
+)
+def check_workflows(  # noqa: C901
+    date_start: Optional[datetime.datetime], date_end: Optional[datetime.datetime]
+) -> None:
+    """Check if selected workflows are in sync with database according to predefined rules."""
+    from reana_commons.k8s.api_client import current_k8s_corev1_api_client
+    from reana_commons.config import REANA_RUNTIME_KUBERNETES_NAMESPACE
+    from reana_server.reana_admin.consumer import CollectingConsumer
+    from kubernetes.client.rest import ApiException
+
+    # filter workflows that need to be checked
+    statuses_to_check = [
+        RunStatus.queued,
+        RunStatus.pending,
+        RunStatus.running,
+        RunStatus.finished,
+        RunStatus.failed,
+    ]
+
+    query_filter = [Workflow.status.in_(statuses_to_check)]
+
+    if date_start:
+        query_filter.append(Workflow.created >= date_start)
+
+    if date_end:
+        query_filter.append(Workflow.created < date_end)
+
+    filtered_workflows = Session.query(Workflow).filter(*query_filter)
+
+    if filtered_workflows.count() == 0:
+        click.secho(
+            "No workflows found. Please, adjust your filter conditions.",
+            fg="red",
+        )
+        return
+
+    click.secho(
+        f"Found {filtered_workflows.count()} workflows. Starting to check them...",
+    )
+
+    # collect messages from workflow-submission queue
+    try:
+        scheduler_messages_collector = CollectingConsumer(
+            queue_name="workflow-submission",
+            key="workflow_id_or_name",
+            values_to_collect=[str(workflow.id_) for workflow in filtered_workflows],
+        )
+    except Exception as error:
+        click.secho(
+            "Error is raised during MessageConsumer initialization. Please, check if arguments are correct.",
+            fg="red",
+        )
+        logging.exception(error)
+        return
+    else:
+        scheduler_messages_collector.run()
+
+    # get snapshot of all present pods
+    try:
+        response = current_k8s_corev1_api_client.list_namespaced_pod(
+            namespace=REANA_RUNTIME_KUBERNETES_NAMESPACE
+        )
+        pods = response.items
+    except ApiException:
+        pods = []
+
+    # validate if workflows are in sync with DB
+    in_sync_workflows = []
+    out_of_sync_workflows = []
+
+    for workflow in filtered_workflows:
```

Comment:
It would be nice to somehow increase the readability of this peace of code (line 803-860). Perhaps breaking it down in small functions would help? Not easy to follow the logic with nested if statements, etc

Reply:
Agree. I will work on that. Any suggestions on how to break it?

Reply:
It's a pity that we can't still use [structural pattern matching](https://docs.python.org/3.10/whatsnew/3.10.html#pep-634-structural-pattern-matching) which would play nicely here, but I think breaking it down in small functions and placing it in some utils file would help. Also we could use dictionary lookups instead of if statements for all the run statuses to point to these functions